### PR TITLE
disttask: support pause/resume for framework

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #46258 

Problem Summary:

### What is changed and how it works?
1. When receive pause command

- Dispatcher:  

pausing state: check all subtasks paused.
paused state: exit the dispatcher.

- Manager: pause and cancel all running subtasks, mark running/pending subtasks as Paused.

2. When receive resume command:

- Dispatcher: check all paused subtasks turn into pending state(not in paused state). 
If yes, turn task into running state.
Else mark all paused subtasks into pending state.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/scheduler/manager.go [hunk 1]
```diff
@@ -181,7 +192,13 @@ func (m *Manager) fetchAndFastCancelTasks(ctx context.Context) {
...
 	tasks = m.filterAlreadyHandlingTasks(tasks)
 	for _, task := range tasks {
-		exist, err := m.taskTable.HasSubtasksInStates(m.id, task.ID, task.Step, proto.TaskStatePending, proto.TaskStateRevertPending)
+		// Need to poll pending/revertPending/Running subtasks.
+		// It's necessary to poll running tasks since manager has possibility of restart.
+		// Then the subtasks are in running state and need to be handled.
+		exist, err := m.taskTable.HasSubtasksInStates(
+			m.id, task.ID, task.Step,
+			proto.TaskStatePending, proto.TaskStateRevertPending, proto.TaskStateRunning)
+
```
Comment (ywqzzy) [inline disttask/framework/scheduler/manager.go:201]: Ignore it

File: disttask/framework/scheduler/manager.go [hunk 2]
```diff
@@ -296,19 +329,24 @@ func (m *Manager) onRunnableTask(ctx context.Context, task *proto.Task) {
 				zap.Int64("task_id", task.ID), zap.Int64("step", task.Step), zap.String("state", task.State))
 			return
 		}
-		if exist, err := m.taskTable.HasSubtasksInStates(m.id, task.ID, task.Step, proto.TaskStatePending, proto.TaskStateRevertPending); err != nil {
+
+		// Considering manager restart scene, scheduler needs to handle running subtasks.
+		if exist, err := m.taskTable.HasSubtasksInStates(
+			m.id, task.ID, task.Step,
+			proto.TaskStatePending, proto.TaskStateRevertPending, proto.TaskStateRunning); err != nil {
```
Comment (ywqzzy) [inline disttask/framework/scheduler/manager.go:336]: Ignore it

File: disttask/framework/scheduler/scheduler.go [hunk 3]
```diff
@@ -170,8 +170,10 @@ func (s *BaseScheduler) run(ctx context.Context, task *proto.Task) error {
 		if err := s.getError(); err != nil {
 			break
 		}
-
-		subtask, err := s.taskTable.GetSubtaskInStates(s.id, task.ID, task.Step, proto.TaskStatePending)
+		// Considering manager restart scene, scheduler needs to handle running subtasks.
+		subtask, err := s.taskTable.GetSubtaskInStates(
+			s.id, task.ID, task.Step,
+			proto.TaskStatePending, proto.TaskStateRunning)
```
Comment (ywqzzy) [inline disttask/framework/scheduler/scheduler.go:176]: Ignore it

File: disttask/framework/scheduler/manager.go [hunk 4]
```diff
@@ -218,6 +235,22 @@ func (m *Manager) onCanceledTasks(_ context.Context, tasks []*proto.Task) {
 	}
 }
 
+// onPausingTasks pauses/cancels the pending/running tasks.
+func (m *Manager) onPausingTasks(tasks []*proto.Task) error {
+	m.mu.RLock()
```
Comment (Benjamin2037) [inline disttask/framework/scheduler/manager.go:240]: Why here add lock? will there concurrent access manager? I thought manager only one for one TiDB.

File: disttask/framework/storage/task_table.go [hunk 5]
```diff
@@ -602,25 +602,24 @@ func (stm *TaskManager) UpdateFailedSchedulerIDs(taskID int64, replaceNodes map[
 	return err
 }
 
-// AddSubTasks add new batch of subtasks.
-func (stm *TaskManager) AddSubTasks(task *proto.Task, subtasks []*proto.Subtask) error {
```
Comment (Benjamin2037) [inline disttask/framework/storage/task_table.go:606]: Why here use PauseSubtasks replace AddSubTasks?

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: disttask/framework/storage/task_table.go (hunk 5 above)
Comment (ywqzzy) [reply disttask/framework/storage/task_table.go:606]: > Why here use PauseSubtasks to replace AddSubTasks?

AddSubtasks is useless, just remove it

File: disttask/framework/scheduler/manager.go (hunk 4 above)
Comment (ywqzzy) [reply disttask/framework/scheduler/manager.go:240]: fetchAndHandleRunnableTasks and fetchAndFastCancelTasks use 2 goroutines  to access handlingTasks.

File: disttask/framework/handle/handle.go [hunk 6]
```diff
@@ -106,18 +106,42 @@ func SubmitAndRunGlobalTask(ctx context.Context, taskKey, taskType string, concu
...
+	if task == nil {
+		logutil.BgLogger().Info("task not exist", zap.String("taskKey", taskKey))
+
+		return nil
+	}
+	return taskManager.CancelGlobalTask(task.ID)
+}
+
+// ResumeTask resumes a task.
+func ResumeTask(taskKey string) error {
```
Comment (zimulala) [inline disttask/framework/handle/handle.go:126]: Do we forget the `PauseTask` interface?

File: disttask/framework/handle/handle.go [hunk 7]
```diff
@@ -106,18 +106,42 @@ func SubmitAndRunGlobalTask(ctx context.Context, taskKey, taskType string, concu
...
+	taskManager, err := storage.GetTaskManager()
+	if err != nil {
+		return err
+	}
+	task, err := taskManager.GetGlobalTaskByKey(taskKey)
```
Comment (zimulala) [inline disttask/framework/handle/handle.go:131]: We check this result in line140(through `found`), so could we remove this check?

File: disttask/framework/storage/task_table.go [hunk 8]
```diff
@@ -602,25 +623,24 @@ func (stm *TaskManager) UpdateFailedSchedulerIDs(taskID int64, replaceNodes map[
...
-	})
-	return err
+// PauseSubtasks update all running/pending subtasks to pasued state.
+func (stm *TaskManager) PauseSubtasks(tidbID string, taskID int64) error {
+	_, err := stm.executeSQLWithNewSession(stm.ctx,
+		`update mysql.tidb_background_subtask set state = "paused" where task_key = %? and state in ("running", "pending") and exec_id = %?`, taskID, tidbID)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (zimulala) [inline disttask/framework/storage/task_table.go:633]: We can return `err` directly.

File: disttask/framework/storage/task_table.go [hunk 9]
```diff
@@ -602,25 +623,24 @@ func (stm *TaskManager) UpdateFailedSchedulerIDs(taskID int64, replaceNodes map[
...
+}
+
+// ResumeSubtasks update all paused subtasks to pending state.
+func (stm *TaskManager) ResumeSubtasks(taskID int64) error {
+	_, err := stm.executeSQLWithNewSession(stm.ctx,
+		`update mysql.tidb_background_subtask set state = "pending", error = null where task_key = %? and state = "paused"`, taskID)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (zimulala) [inline disttask/framework/storage/task_table.go:643]: Ditto.

File: disttask/framework/storage/task_table.go [hunk 10]
```diff
@@ -352,10 +352,31 @@ func (stm *TaskManager) UpdateErrorToSubtask(tidbID string, taskID int64, err er
...
+			} else {
+				err = stdErr
+			}
+		}
+		logutil.BgLogger().Info(fmt.Sprintf("subTask: %v\n", row2SubTask(r)), zap.Error(err))
+	}
+}
+
+// PrintHistorySubtaskInfo log the history subtask info by taskKey. Only used for UT.
+func (stm *TaskManager) PrintHistorySubtaskInfo(_ int64) {
```
Comment (zimulala) [inline disttask/framework/storage/task_table.go:377]: Could we use a unified function for `PrintSubtaskInfo` and `PrintHistorySubtaskInfo`?

File: disttask/framework/handle/handle.go (hunk 7 above)
Comment (ywqzzy) [reply disttask/framework/handle/handle.go:131]: > We check this result in line140(through `found`), so could we remove this check?

I will change the resumeTask(int64) to resumeTask(string)

**Comment by zimulala** [comment]
Please fix the CI @ywqzzy

**Pull request merged**